tool/sample-synthetic-data.ts

```diff
@@ -0,0 +1,511 @@
+        ];
+    case 'reverse_property':
+    case 'reverse_property_true':
+    case 'reverse_property_false':
+        return [
+            example(`${queryCanonical} ${predicate}?`)
+        ];
+    case 'adjective':
+    case 'adjective_true':
+    case 'adjective_false':
+        return [
+            example(`${queryCanonical} ${predicate} .`),
+        ];
+    case 'reverse_verb':
+        return [
+            example(`${queryCanonical} ${predicate}`)
+        ];
+    default:
+        return [];
+    }
+    
+}
+
+function generateFilterAst(device : string, 
+                           func : string, 
+                           property : string, 
+                           operator : string, 
+                           value : any) : Ast.Program {
+    const invocation = new Ast.InvocationExpression(
+        null, 
+        new Ast.Invocation(null, new Ast.DeviceSelector(null, device, null, null), func, [], null),
+        null
+    );
+    const filter = new Ast.AtomBooleanExpression(
+        null,
+        property, 
+        operator, 
+        value, // 
+        null
+    );
+    const filtered = new Ast.FilterExpression(
+        null, 
+        invocation, 
+        filter, 
+        null
+    );
+    const statement = new Ast.ExpressionStatement(
+        null,
+        new Ast.ChainExpression(null, [filtered], null)
+    );
+    return new Ast.Program(
+        null,
+        [], 
+        [],
+        [statement],
+        {}
+    );
+}
+
+function generateProjectionAst(device : string, func : string, property : string) : Ast.Program {
+    const invocation = new Ast.InvocationExpression(
+        null, 
+        new Ast.Invocation(null, new Ast.DeviceSelector(null, device, null, null), func, [], null),
+        null
+    );
+    const projection = new Ast.ProjectionExpression(
+        null,
+        invocation, 
+        [property],
+        [], 
+        [], 
+        null
+    );
+    const statement = new Ast.ExpressionStatement(
+        null,
+        new Ast.ChainExpression(null, [projection], null)
+    );
+    return new Ast.Program(
+        null,
+        [], 
+        [],
+        [statement],
+        {}
+    );
+}
+
+function generateActionAst(device : string, 
+                           func : string, 
+                           property : string, 
+                           operator : string, 
+                           value : any) : Ast.Program {
+    const invocation = new Ast.InvocationExpression(
+        null, 
+        new Ast.Invocation(null, new Ast.DeviceSelector(null, device, null, null), func, [], null),
+        null
+    );
+    const filter = new Ast.AtomBooleanExpression(
+        null,
+        property, 
+        operator, 
+        value, // 
+        null
+    );
+    const filtered = new Ast.FilterExpression(
+        null, 
+        invocation, 
+        filter, 
+        null
+    );
+    const inputParam = new Ast.InputParam(null, property, value);
```

Comment:
It's either an output (filter) or input, it can't be both. I think we can assume for action, it's always input param